### PR TITLE
Implements new plugin based on npmUpdate

### PR DESCRIPTION
**Pull request by mathieuleclaire**
Here is an implementation of the plugin disscused in #85. The idea is to use npmUpdate to fetch js files without using webpack. 

I introduced a new case class to define which files are required in the npm archives. I like the fact that needed js are defined in the build.sbt and not in @JSImport("module/foo/bar"). And actually without webpack, I think I have no other option. 

It is called NpmDepsPlugin (not very sexy, it can be updated :)). Typically in a project it is used this way:

```scala
npmDeps in Compile += Dep("ace-builds", "1.2.9", List("ace.js", "mode-scala.js", "theme-github.js"))
```
Then a ```dependencyFile``` key enables to get the file containing all the npm js libs are merged.
That's it !

Typically, you can have this facade in a project:
```scala
lazy val ace = project.in(file("ace")) enablePlugins (NpmDepsPlugin) settings (defaultSettings) settings (
  scalaJsDom,
  npmDeps in Compile += Dep("ace-builds", aceVersion, List("ace.js", "mode-scala.js", "theme-github.js"))
  )
```
And then:
```scala
lazy val demo = project.in(file("demo")) enablePlugins (NpmDepsPlugin) settings (defaultSettings) settings(
  libraryDependencies += "com.lihaoyi" %%% "scalarx" % rxVersion,
  libraryDependencies += "com.lihaoyi" %%% "sourcecode" % sourceCodeVersion,
  libraryDependencies += "com.github.karasiq" %%% "scalajs-marked" % "1.0.2",
  runDemo := {

    val demoTarget = target.value
    val demoResource = (resourceDirectory in Compile).value

    val jsTarget = demoTarget / "js"
    val demoJS = (fullOptJS in Compile).value

    IO.copyFile(demoJS.data, jsTarget / "demo.js")
    IO.copyFile(dependencyFile.value, jsTarget / "deps.js")

    IO.copyFile(demoResource / "bootstrap-native.html", demoTarget / "index.html")
  }
) dependsOn(ace)
```
Of cours, new npmDeps can be added in the demo project.

It also work if ace is published as a library.

Some points may be upgraded:
- [x] npmUpdate  have been copy / pasted / modified. Peharps, something more generic could be done to have only one implementation to both ```ScalaJSPlugin``` and ```NpmDepsPlugin```
- [x] Keep only ```npmDependencies``` / ```npmUpdate``` stuff in the ```NpmDepsPlugin``` and start a new plugin project inheriting from ```NpmDepsPlugin```  and implementing the jsDependencies part.

### CODE REVIEW

**Comment by sjrd** [comment]
Why doesn't `ScalaJSBundlerPlugin` depend on the new `NpmDepsPlugin`? I would hope that `NpmDepsPlugin` can factor out the npm-related stuff from `ScalaJSBundlerPlugin`, not just be completely independent of it.

**Comment by sjrd** [comment]
In the same spirit, why add a distinct `npmDeps` instead of reusing `npmDependencies`? In any case, having both is super confusing.

**Comment by mathieuleclaire** [comment]
I will try to depends on ScalaJSBundlerPlugin, since I have something working well. It is what I tried first but it rendered things too complex at this stage.

I agree that npmDeps is a new key, but I am not sure I can use @JImport instructions to fetch several js in the same module. In my example, I get 3 files in the ace-builds module. This way, it is easy to define which js you want when you define your build.

**Comment by sjrd** [comment]
> In my example, I get 3 files in the ace-builds module. This way, it is easy to define which js you want when you define your build.

This concern should be separated from the `npmDependencies`. This is a way to tell `resolvedJSDependencies` to take specific modules from `node_modules/` and *execute* them in the context of `run` and `test`. This is very specific. You should have a distinct taskKey for that, that only affects `resolvedJSDependencies`, while `npmDependencies` only affects what gets *installed* (and not necessarily *executed*).

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      }
+
+      val resolvedDependencies = jss.map { f =>
+        ResolvedJSDependency.minimal(FileVirtualJSFile(f))
+      }
+
+      prev.map(_ ++ resolvedDependencies)
+    },
+
+    dependencyFile := (packageMinifiedJSDependencies in Compile).value,
```
Comment (sjrd) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: This is defined, but never *read*. Looks like dead code.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  ) ++ inConfig(Compile)(perConfigSettings)
+
+
+  private lazy val perConfigSettings: Seq[Def.Setting[_]] = Seq(
+    allNpmDeps := NpmDeps.collectFromClasspath((fullClasspath in Compile).value),
+
+    npmUpdate := {
+      val log = streams.value.log
+      val targetDir = (crossTarget in npmUpdate).value
+      val jsResources = scalaJSNativeLibraries.value.data
```
Comment (sjrd) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: Don't introduce new code depending on `scalaJSNativeLibraries`. It's gone in Scala.js 1.x.

**Comment by mathieuleclaire** [comment]
so something like: 
```scala 
npmDependencies += "ace-builds"->  "1.2.9"
otherKey += "ace-builds"-> List("ace.js", "mode-scala.js", "theme-github.js")
``` 
? It's twice as many lines to write

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 1 above)
Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: dependencyFile is the Key to get the packagedMinifiedJSDependencies. I use it in my demo to build deps.js.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 2 above)
Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: It is copy/pasted from ScalaJSBundlerPlugin. How to replace this ?

**Comment by sjrd** [comment]
Well, if you really want, you can encode everything in `otherKey`, and derive `npmDependencies` from it. For example, in the plugin:
```scala
npmDependencies ++= otherKey.value.map(_._1)
```
and in the build:
```scala
otherKey += ("ace-builds", "1.2.9") -> List(...)
```
Note that this even highlights that `otherKey` could be part of a *third* plugin, which depends on `NpmDepsPlugin`, maybe called `NpmDepsExecutionFilesPlugin`.

In any case, it leaves the opportunity to use `NpmDepsPlugin` with `npmDependencies` only, without altering the execution files (which is what `ScalaJSBundlerPlugin` wants).

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 1 above)
Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: Why don't you use `(packageMinifiedJSDependencies in Compile).value` in your demo instead, then?

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 2 above)
Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: Uh, well, if you *cut*-and-paste it from `ScalaJSBundlerPlugin` (which you should if `NpmDepsPlugin` factors out stuff from `ScalaJSBundlerPlugin` instead of living beside it), then keep it, as the PR doesn't make the situation worse.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 1 above)
Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: To render the API smarter. Here, you only have to call ```dependencyFile.value```, and not both:
```scala
skip in packageJSDependencies := false
...
packageMinifiedJSDependencies in Compile).value
```

Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: In the current definition of the plugin, `dependencyFile` does not influence `skip in packageJSDependencies`, so anyway you need two "things":
```scala
skip in packageJSDependencies := false
...
dependencyFile.file
```
I do not see the added value.

Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: Yes, but it is hidden in the NpmDepsPlugin project settings. So that, you only have one line to write in your project definition and not 2.

Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: My comment here does not suggest removing `skip in packageJSDependencies := false` from the plugin. If you just remove the `dependencyFile` from the sbt plugin, then the only thing that changes in the demo is to replace `dependencyFile.value` by `(packageMinifiedJSDependencies in Compile).value`. It is still one line in the project definition.

Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: You're right. On the other hand, isn't it a little bit confusing since these two lines are still use together in scalajs ?

**Comment by mathieuleclaire** [comment]
But, npmDependencies is actually not used in NpmDepsPlugin

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 1 above)
Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: > On the other hand, isn't it a little bit confusing since these two lines are still use together in scalajs ?

The whole settings around `jsDependencies` is confusing because it's mostly broken, and that's why it's been extracted away so that people who don't need it can ignore it.

Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: No. I recommend that `NpmDepsPlugins` only takes care of `npmDependencies`, `npmUpdate`, and immediate friends that do not depend on `jsDependencies` (e.g., `useYarn`, `packageJson`).

Then, you could start a new project that links `NpmDepsPlugins` with `jsDependencies`, `resolvedJSDependencies`, and friends (e.g., `packageMinifiedJSDependencies).

`NpmDepsPlugin` would be the common denominator of `ScalaJSBundlerPlugin` and your other plugin.

Less magic, less responsibilities per plugin -> win/win for everyone, because everyone can choose how much they need.

Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: I think it should be a very different project, so that the scalajs-bundler repo can eventually be rid of the dependency on (what's been moved to) sbt-jsdependencies.

That said, you can probably *develop* it first in the same repo so that you can iterate more quickly, but then PR only the part for `NpmDepsPlugin`.

**Comment by mathieuleclaire** [comment]
So, I kept NpmDepsPlugin as a generic plugin implementing npmUpdate and I moved my particular keys in a new separate plugin depending on NpmDepsPlugin: https://github.com/ISCPIF/scalajs-execnpm. What do you think ?

File: build.sbt [hunk 3]
```diff
@@ -8,6 +8,7 @@ val `sbt-scalajs-bundler` =
     .settings(commonSettings)
     .settings(
       sbtPlugin := true,
+      version := "0.11.0",
```
Comment (sjrd) [inline build.sbt]: Not sure how @julienrf manages versions, but this should probably not be here.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    TaskKey[BundlerFile.PackageJson]("scalaJSBundlerPackageJson",
+      "Write a package.json file defining the NPM dependencies of project",
+      KeyRanks.Invisible
+    )
+
+  override lazy val projectSettings =
+    perScalaJSStageSettings(fullOptJS) ++ perScalaJSStageSettings(fastOptJS)
+
+  protected def perScalaJSStageSettings(stage: TaskKey[Attributed[File]]): Seq[Def.Setting[_]] = Seq(
+    useYarn := false
```
Comment (sjrd) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: In `ScalaJSBundlerPlugin`, `useYarn` used to be set in the project scope, directly in `projectSettings`. I think that's also where it should be in `NpmDepsPlugin`.

Speaking of which, the `useYarn := false` should be removed from `ScalaJSBundlerPlugin.projectSettings`, since it's redundant with the assignment in `NpmDepsPlugin.projectSettings`.

Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: In fact, you should inline the contents of your current `persScalaJSStageSettings` inside `projectSettings`, since you're not using `stage` at all here (which makes sense). So currently, you're just duplicating all these settings by calling `perScalaJSStageSettings(fullOptJS) ++ perScalaJSStageSettings(fastOptJS)`, for no reason.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  ) ++ inConfig(Compile)(perConfigSettings)
+
+
+  private lazy val perConfigSettings: Seq[Def.Setting[_]] = Seq(
+    npmUpdate := {
+      val log = streams.value.log
+      val targetDir = (crossTarget in npmUpdate).value
+      val jsResources = scalaJSNativeLibraries.value.data
+      val packageJsonFile = scalaJSBundlerPackageJson.value
```
Comment (sjrd) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: You're using `scalaJSBundlerPackageJson.value` from `NpmDepsPlugin`, but `NpmDepsPlugin` does not assign any value to that task. I'm not sure how `NpmDepsPlugin` can be used on its own under those circumstances.

You should probably add a scripted test with a project that only uses `NpmDepsPlugin` (but not `ScalaJSBundlerPlugin`), and does something meaningful with it. You might uncover other similar issues.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      "Write a package.json file defining the NPM dependencies of project",
+      KeyRanks.Invisible
+    )
+
+  override lazy val projectSettings =
+    perScalaJSStageSettings(fullOptJS) ++ perScalaJSStageSettings(fastOptJS)
+
+  protected def perScalaJSStageSettings(stage: TaskKey[Attributed[File]]): Seq[Def.Setting[_]] = Seq(
+    useYarn := false
+  ) ++ inConfig(Compile)(perConfigSettings)
```
Comment (sjrd) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: You also need `inConfig(Test)(perConfigSettings)` here, otherwise `npmUpdate` cannot be used in the `Test` configuration.

File: build.sbt (hunk 3 above)
Comment (mathieuleclaire) [reply build.sbt]: Yes, I have to correct this, it was for my tests.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 5 above)
Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: scalaJSBundlerPackageJson is set in ScalaJSBundlerPlugin ```perConfigSettings``` and set it in my ExecNpmPlugin ```perConfigSettings```

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 6 above)
Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: I commited my modifications except for the scalaJSBundlerPackageJson. Sadly I was off the previous 15 days ;) Do you think you and @julienrf that a 0.10.1 could be published with this PR, once it will be OK ?

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 5 above)
Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: That means that `NpmDepsPlugin` is not useful on its own, so it is probably not a good delineation of an abstraction. `NpmDepsPlugin` should define `scalaJSBundlerPackageJson` itself (and it should be called `npmDepsPackageJson`), in a way that is generic enough to be reused.

From the way it is defined in `ScalaJSBundlerPlugin`, I don't see why the definition there cannot be factored out (almost) as is into `NpmDepsPlugin`. The only bundler-specific things there are `(version in webpack).value` and `(version in startWebpackDevServer).value`, which means that those two things need an additional hook in `NpmDepsPlugin`. Other keys such as `npmDependencies` can be extracted into `NpmDepsPlugin` as well. It makes sense. Then in your plugin `ExecNpmPlugin` you could automatically derive `npmDependencies` from your `allNpmDeps`, and leave the other ones alone since they would have sensible defaults. Doing it this way will result in even less code in `ExecNpmPlugin`, which kind of proves that it would provide a better abstraction.

Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: I didn't define ```scalaJSBundlerPackageJson``` in NpmDepsPlugin because I don't want it to depend on ```webpack```. I redefine ```scalaJSBundlerPackageJson``` with fake values in my ```ExecNpmPlugin```. So, what should I do ? Define two new settingsKey for both versions at the ```NpmDepsPlugin``` level (so that their values are set in ```scalaJSBundlerPlugin``` with the webpack version values and with fake ones in the ```ExecNpmPlugin```) ?

Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: > Define two new settingsKey for both versions at the NpmDepsPlugin level (so that their values are set in scalaJSBundlerPlugin with the webpack version values and with fake ones in the ExecNpmPlugin) ?

Yes, I think that's what is needed, here.

Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: Or perhaps better, remove `webpackVersion` and `webpackDevServerVersion` from the APIs of `PackageJson` altogether. In that case, `ScalaJSBundlerPlugin` would automatically add the relevant pairs to the `npmDevDependencies`, as is currently done [here](https://github.com/scalacenter/scalajs-bundler/blob/e089cf44e0066e484a8799cc35019ea8a406ff14/sbt-scalajs-bundler/src/main/scala/scalajsbundler/PackageJson.scala#L42-L59).

So `ScalaJSBundlerPlugin.perConfigSettings` would contain something like
```scala
npmDevDependencies ++= {
  val webpackVersion := (version in webpack).value
  val webpackDevServerVersion := (version in startWebpackDevServer).value

   val sourceMapLoaderVersion = 
      NpmPackage(webpackVersion).major match {
        case Some(1) | Some(2) => "0.1.5"
        case Some(3) => "0.2.1"
        case Some(x) => sys.error(s"Unsupported webpack major version $x")
        case None => sys.error("No webpack version defined")
      }

    npmDevDependencies.value ++ Seq(
      "webpack" -> webpackVersion,
      "webpack-dev-server" -> webpackDevServerVersion,
      "concat-with-sourcemaps" -> "1.0.4", // Used by the reload workflow
      "source-map-loader" -> sourceMapLoaderVersion // Used by webpack when emitSourceMaps is enabled
    )
}
```
(more or less)

and then `NpmDepsPlugin` wouldn't have to even know about `webpackVersion` or `webpackDevServerVersion`. And you wouldn't need any fake version for those in your plugin. :)

Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: Yes, but I still have to define ```npmDependencies``` and stuff in NpmDepsPlugin, which does not really make sense. 
Actually I don't need to have ```devDependencies``` to be written at all in my package.json. Why not having a ```writePackageJson``` function with optional ```devDependencies``` arguments(```npmDependencies.value```, ```npmDevDependencies.value```, ```npmResolutions.value```,```additionalNpmConfig.value```, and both webpack versions) ?

Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: Why would `npmDependencies` not make sense in `NpmDepsPlugin`!? On the contrary, it makes *perfect* sense. It's even the name of the plugin: *NpmDeps*Plugin.

And the other arguments are already "optional" in the sense that they are collections, and their default values are to be empty. No need to make them more optional than that.

Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: > Why would npmDependencies not make sense in NpmDepsPlugin!?

Deps is not the same than ```Dependencies```, it is a collection of ``` case class Dep(module: String, version: String, jsFiles: List[String])```. But now it has been moved in ```ExecNpmPlugin```. So ```NpmDepsPlugin``` should be renamed to avoid confusion. Something like NpmUpdatePlugin.

And, you right, adding Options is ugly. 
Well, it all about 3 lines of json to be added or not ! :)

Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: I splited the write function of package.json into devDependencies and depencies parts. The ```PackageJson.write``` method is still called from ```ScalaJSBundlerPlugin``` and ```PackageJson.writeDepsOnly``` is called from my ```ExecNpmPlugin```. This way, no more fake values and partial fake package.json file are generated.

**Comment by sjrd** [comment]
I don't understand where we're going anymore. The way it's done currently, `NpmUpdatePlugin` is *just a function*, and would be advantageously replaced by precisely that: just a standalone `def`, no plugin whatsoever.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala (hunk 5 above)
Comment (sjrd) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala]: > Deps is not the same than Dependencies, it is a collection of case class Dep(module: String, version: String, jsFiles: List[String]).

My point all along has been that you can *derive* `npmDependencies` from what your `Dep`s. And that therefore, `NpmDepsPlugin` (which should perhaps be called `NpmDependenciesPlugin`, to avoid naming clashes) can be implemented in terms of `npmDependencies`, and that your plugin can automatically derive `npmDependencies` in terms of what you call `npmDeps`.

If we're not going at least *that* far in factoring out things out of `ScalaJSBundlerPlugin`, then `NpmUpdatePlugin` is really *just* a `def`, and should not be an sbt plugin at all.

**Comment by mathieuleclaire** [comment]
What ```def``` ? The ```npmUpdate``` ? I use ```npmUpdate``` and ```writeDepsOnly``` ? I don't understand how could use it as an function ?

**Comment by sjrd** [comment]
`npmUpdate`. It really is just a tiny function, with some inputs and some side effects and outputs. If the inputs do not receive sensible defaults from other parts of the sbt plugin as other settings and tasks, then the sbt plugin is not providing any feature beside having a `def npmUpdate(inputs): Outputs`.

Note that it is entirely possible that it is the right thing to do for your use case. In that case, it may be better to revert to the original state, and just expose this `def npmUpdate` as a part of the non-sbt API, which you can call from your plugin.

**Comment by mathieuleclaire** [comment]
You are right ! I removed the ```NpmUpdatePlugin``` and I defined npmUpdate in a ```NpmUpdateTasks``` object. All in all, I have two new function definitions: ```npmUpdate``` and ```writeOnlyDepsPackageJson```. My ```ExecNpmPlugin``` works fine with these 2 new functions. Is it now acceptable for you ?

**Comment by mathieuleclaire** [comment]
@julienrf : is it fine for you ?

**Comment by julienrf** [comment]
@mathieuleclaire A lot of discussion has happened around this PR, is the PR description up to date? Can you guide us to review the PR?

**Comment by mathieuleclaire** [comment]
Not exactly. @sjrd finally said my plugin should be external to scalajs-bundler. But I need to call 2 adapded functions: 
- npmUpdate in a separate object: https://github.com/mathieuleclaire/scalajs-bundler/blob/0862712bc8714ac29d1a9ba260450a79dab080b0/sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmUpdateTasks.scala#L15:L49 and 
- writeOnlyDepsPackageJson: https://github.com/mathieuleclaire/scalajs-bundler/blob/0862712bc8714ac29d1a9ba260450a79dab080b0/sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/PackageJsonTasks.scala#L79:L111

It has no effect on the scalajs-bundler behaviour (just calling npmUpdate in a separate object) and it is very helpfull for me since I can use npmUpdate out of the scalajs-bundler plugin context.

**Comment by julienrf** [comment]
What do you think of the initial idea of having a plugin on which `ScalaJSBundlerPlugin` depends? That plugin would define the `npmDependencies` and `npmDevDependencies` keys and expose a task to write a `package.json` file. Then, `ScalaJSBundlerPlugin` would add it’s own entries to the `npmDevDependencies` key. And we would reuse the same method to generate the `package.json` file.

**Comment by mathieuleclaire** [comment]
Well, we are talking since months about this issue. I changed my implementations few times according to what you wanted (create a third plugin inside scalajs-bundler, then implement separate functions, then implement an external plugin). My only goal was not to duplicate scalajs-bundler functions. Anyway. 

Could you at least accept [the npmUpdate function in a separate object](https://github.com/mathieuleclaire/scalajs-bundler/blob/0862712bc8714ac29d1a9ba260450a79dab080b0/sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmUpdateTasks.scala#L15:L49), so that it can be used in external projects ?

Thanks

**Comment by julienrf** [comment]
Yes.

**Comment by mathieuleclaire** [comment]
How do we proceed ?

**Comment by julienrf** [comment]
It’s up to you, you can update that PR or start a new one, it doesn’t really matter for me.

**Comment by mathieuleclaire** [comment]
Here we are !

**Comment by julienrf** [comment]
Sorry to bother you again but can you remove the changes you’ve applied to code style only, and squash your commits into a single one?

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmUpdateTasks.scala [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import scalajsbundler.BundlerFile.PackageJson
+
+object NpmUpdateTasks {
+
+  def npmUpdate(targetDir: File,
+                packageJsonFile: File,
+                useYarn: Boolean,
+                jsResources: Seq[VirtualJSFile with RelativeVirtualFile],
+                streams: Keys.TaskStreams
+                ) = {
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmUpdateTasks.scala]: Would you mind adding some scaladoc?

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala [hunk 8]
```diff
@@ -47,7 +47,7 @@ import scalajsbundler.util.JSON
   * Version of webpack-dev-server to use.
   *
   * {{{
-  *   version in startWebpackDevServer := "2.7.1"
+  *   version in startWebpackDevServer := "2.11.1"
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: Why this change?

Comment (mathieuleclaire) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: No idea :). My mistake.

**Comment by mathieuleclaire** [comment]
How could I squash all my commits into a single one ? By reverting and recommiting ?

**Comment by julienrf** [comment]
Nevermind, I can squash them by myself if you are not familiar with that process. In that case, just make sure that the PR only contains the minimum amount of changes (ie no code style changes).

**Comment by mathieuleclaire** [comment]
I hope it is fine now :) !

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala [hunk 9]
```diff
@@ -324,6 +324,18 @@ object ScalaJSBundlerPlugin extends AutoPlugin {
...
       taskKey[Seq[File]]("Files that trigger webpack launch")
 
+    /**
+      * Additional arguments to webpack
+      *
+      * Defaults to an empty list.
+      *
+      * @group settings
+      */
+    val webpackExtraArgs = SettingKey[Seq[String]](
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: This key already exists in master. So, you should rebase your work on top of master.

In the end, it’s probably simpler to start a fresh PR on top of master.
